### PR TITLE
namespace controller: use contextual logging

### PR DESCRIPTION
**Pull request by yangjunmyfm192085**
#### What type of PR is this?
/kind feature

#### Which issue(s) this PR fixes:

Fixes #
Related-to: https://github.com/kubernetes/enhancements/issues/3077
#### Does this PR introduce a user-facing change?

```release-note
Migrated the namespace controller (within `kube-controller-manager`) to support [contextual logging](https://k8s.io/docs/concepts/cluster-administration/system-logs/#contextual-logging).
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Review by yangjunmyfm192085 — COMMENTED** [review]
/retest

**Review by yangjunmyfm192085 — COMMENTED** [review]
/sig instrumentation
/wg structured-logging

**Review by pohly — CHANGES_REQUESTED** [review]

File: cmd/kube-controller-manager/app/core.go
Comment (yangjunmyfm192085) [reply cmd/kube-controller-manager/app/core.go:445]: done

File: pkg/controller/namespace/deletion/namespaced_resources_deleter.go
Comment (yangjunmyfm192085) [reply pkg/controller/namespace/deletion/namespaced_resources_deleter.go]: done

**Comment by yangjunmyfm192085** [comment]
> There's one valid compile issue:
> 
> ```
> test/integration/namespace/ns_conditions_test.go:132:22: cannot use 5 (constant of type int) as "context".Context value in argument to nsController.Run: int does not implement "context".Context (missing method Deadline) (typecheck)
> 	go nsController.Run(5, stopCh)
> ```

But the strange thing is that the code is not `go nsController.Run(5, stopCh)`, but `go nsController.Run(ctx, 5)`

**Comment by pohly** [comment]
Strange failures like this can occur because what gets tested is the combination of the PR and the current master branch. In this case, the master branch has another `go nsController.Run` invocation. After merging:

```
$ git grep nsController.Run
staging/src/k8s.io/apiserver/pkg/admission/plugin/validatingadmissionpolicy/controller.go:              c.policyDefinitionsController.Run(ctx)
test/integration/namespace/ns_conditions_test.go:       go nsController.Run(ctx, 5)
test/integration/namespace/ns_conditions_test.go:       go nsController.Run(5, stopCh)
```

You need to rebase, then fix that other call.

**Comment by yangjunmyfm192085** [comment]
> Strange failures like this can occur because what gets tested is the combination of the PR and the current master branch. In this case, the master branch has another `go nsController.Run` invocation. After merging:
> 
> ```
> $ git grep nsController.Run
> staging/src/k8s.io/apiserver/pkg/admission/plugin/validatingadmissionpolicy/controller.go:              c.policyDefinitionsController.Run(ctx)
> test/integration/namespace/ns_conditions_test.go:       go nsController.Run(ctx, 5)
> test/integration/namespace/ns_conditions_test.go:       go nsController.Run(5, stopCh)
> ```
> 
> You need to rebase, then fix that other call.

thanks, Let me fix it

File: pkg/controller/namespace/deletion/namespaced_resources_deleter.go
Comment (pohly) [inline pkg/controller/namespace/deletion/namespaced_resources_deleter.go:315]: Please replace "nameSpace " -> "namespace" in the entire PR.

File: pkg/controller/namespace/deletion/namespaced_resources_deleter.go
Comment (pohly) [inline pkg/controller/namespace/deletion/namespaced_resources_deleter.go:563]: You can add `err := utilerrors.NewAggregate(errs)` before this and then use `err` here and in the return in the next line.

**Review by pohly — CHANGES_REQUESTED** [review]

File: pkg/controller/namespace/deletion/namespaced_resources_deleter.go
Comment (yangjunmyfm192085) [reply pkg/controller/namespace/deletion/namespaced_resources_deleter.go]: done

File: pkg/controller/namespace/deletion/namespaced_resources_deleter.go
Comment (yangjunmyfm192085) [reply pkg/controller/namespace/deletion/namespaced_resources_deleter.go]: done

**Comment by sftim** [comment]
Suggestion for release note:
```markdown
Migrated the namespace controller (within `kube-controller-manager`) to support [contextual logging](https://k8s.io/docs/concepts/cluster-administration/system-logs/#contextual-logging).

No user action required: the kube-controller-manager may now exit with code 1
```

Is that right, though? What's the actual thing end users need to know about the exit code(s) from `kube-controller-manager`?

**Comment by yangjunmyfm192085** [comment]
> 

Users need to be informed that the exit code has been changed from 255 to 1

> Suggestion for release note:
> 
> ```gfm
> Migrated the namespace controller (within `kube-controller-manager`) to support [contextual logging](https://k8s.io/docs/concepts/cluster-administration/system-logs/#contextual-logging).
> 
> No user action required: the kube-controller-manager may now exit with code 1
> ```
> 
> Is that right, though? What's the actual thing end users need to know about the exit code(s) from `kube-controller-manager`?

already updated, thanks. Here is the description Exit code changed from 255 to 1

**Review by pohly — APPROVED** [review]
/lgtm

**Comment by pohly** [comment]
/assign @soltysh 

For approval.

File: pkg/controller/namespace/namespace_controller.go
Comment (soltysh) [inline pkg/controller/namespace/namespace_controller.go:137]: Any particular reason for this approach rather than accepting context as a parameter to worker? That way the function will be simple than what is proposed here.

**Review by soltysh — COMMENTED** [review]
/hold 
until we get the answer to the question above

File: pkg/controller/namespace/namespace_controller.go
Comment (pohly) [reply pkg/controller/namespace/namespace_controller.go:137]: I agree, that should also work.

Comment (yangjunmyfm192085) [reply pkg/controller/namespace/namespace_controller.go]: ok, I have updated it

**Comment by pohly** [comment]
/retest

Looks okay to me now.

@soltysh ?

File: cmd/kube-controller-manager/app/core.go
Comment (helayoty) [inline cmd/kube-controller-manager/app/core.go:459]:
```suggestion
	ctx = klog.NewContext(ctx, klog.LoggerWithName(klog.FromContext(ctx), "namespace-controller"))
```

File: pkg/controller/namespace/deletion/namespaced_resources_deleter.go
Comment (helayoty) [inline pkg/controller/namespace/deletion/namespaced_resources_deleter.go:182]:
```suggestion
				logger.V(6).Info("Skipping resource because it cannot be deleted.", "resource",  gvr)
```

File: pkg/controller/namespace/deletion/namespaced_resources_deleter.go
Comment (helayoty) [inline pkg/controller/namespace/deletion/namespaced_resources_deleter.go:315]:
```suggestion
		logger.V(5).Info("Namespace controller - deleteCollection ignored since not supported -", "namespace", namespace, "groupVersionResource", gvr)
```

File: pkg/controller/namespace/deletion/namespaced_resources_deleter.go
Comment (helayoty) [inline pkg/controller/namespace/deletion/namespaced_resources_deleter.go:311]:
```suggestion
	logger.V(5).Info("Namespace controller - deleteCollection -", "namespace", namespace, "groupVersionResource", gvr)
```

File: pkg/controller/namespace/deletion/namespaced_resources_deleter.go
Comment (helayoty) [inline pkg/controller/namespace/deletion/namespaced_resources_deleter.go:111]:
```suggestion
	klog.FromContext(ctx).V(5).Info("Namespace controller - syncNamespace -", "namespace", namespace.Name, "finalizerToken", d.finalizerToken)
```

Comment (helayoty) [inline pkg/controller/namespace/deletion/namespaced_resources_deleter.go:111]:
```suggestion
	klog.FromContext(ctx).V(5).Info("Namespace controller - syncNamespace -", "namespace", namespace.Name, "finalizerToken", d.finalizerToken)
```

File: pkg/controller/namespace/deletion/namespaced_resources_deleter.go
Comment (helayoty) [inline pkg/controller/namespace/deletion/namespaced_resources_deleter.go:336]:
```suggestion
		logger.V(5).Info("Namespace controller - deleteCollection not supported -", "namespace", namespace, "groupVersionResource", gvr)
```

File: pkg/controller/namespace/deletion/namespaced_resources_deleter.go
Comment (helayoty) [inline pkg/controller/namespace/deletion/namespaced_resources_deleter.go:340]:
```suggestion
	logger.V(5).Info("Namespace controller - deleteCollection unexpected error -", "namespace", namespace, "groupVersionResource", gvr, "err", err)
```

File: pkg/controller/namespace/deletion/namespaced_resources_deleter.go
Comment (helayoty) [inline pkg/controller/namespace/deletion/namespaced_resources_deleter.go:356]:
```suggestion
		logger.V(5).Info("Namespace controller - listCollection ignored since not supported -", "namespace", namespace, "groupVersionResource", gvr)
```

File: pkg/controller/namespace/deletion/namespaced_resources_deleter.go
Comment (helayoty) [inline pkg/controller/namespace/deletion/namespaced_resources_deleter.go:380]:
```suggestion
	klog.FromContext(ctx).V(5).Info("Namespace controller - deleteEachItem -", "namespace", namespace, "groupVersionResource", gvr)
```

File: pkg/controller/namespace/deletion/namespaced_resources_deleter.go
Comment (helayoty) [inline pkg/controller/namespace/deletion/namespaced_resources_deleter.go:417]:
```suggestion
	logger.V(5).Info("Namespace controller - deleteAllContentForGroupVersionResource -", "namespace", namespace, "groupVersionResource", gvr)
```

File: pkg/controller/namespace/deletion/namespaced_resources_deleter.go
Comment (helayoty) [inline pkg/controller/namespace/deletion/namespaced_resources_deleter.go:422]:
```suggestion
		logger.V(5).Info("Namespace controller - deleteAllContentForGroupVersionResource - unable to estimate", "namespace", namespace, "groupVersionResource", gvr, "err", err)
```

Comment (helayoty) [reply pkg/controller/namespace/deletion/namespaced_resources_deleter.go:422]: @soltysh Should we change this log to Error instead of Info?

File: pkg/controller/namespace/deletion/namespaced_resources_deleter.go
Comment (helayoty) [inline pkg/controller/namespace/deletion/namespaced_resources_deleter.go:425]:
```suggestion
	logger.V(5).Info("Namespace controller - deleteAllContentForGroupVersionResource - estimate", "namespace", namespace, "groupVersionResource", gvr, "estimate", estimate)
```

File: pkg/controller/namespace/deletion/namespaced_resources_deleter.go
Comment (helayoty) [inline pkg/controller/namespace/deletion/namespaced_resources_deleter.go:443]:
```suggestion
	logger.V(5).Info("Namespace controller - deleteAllContentForGroupVersionResource - checking for no more items in namespace", namespace, "groupVersionResource", gvr)
```

File: pkg/controller/namespace/deletion/namespaced_resources_deleter.go
Comment (helayoty) [inline pkg/controller/namespace/deletion/namespaced_resources_deleter.go:446]:
```suggestion
		logger.V(5).Info("Namespace controller - deleteAllContentForGroupVersionResource - error verifying no items in namespace", namespace, "groupVersionResource", gvr, "err", err)
```

File: pkg/controller/namespace/deletion/namespaced_resources_deleter.go
Comment (helayoty) [inline pkg/controller/namespace/deletion/namespaced_resources_deleter.go:452]:
```suggestion
	logger.V(5).Info("Namespace controller - deleteAllContentForGroupVersionResource - items remaining -", "namespace", namespace, "groupVersionResource", gvr, "items", len(unstructuredList.Items))
```

Comment (helayoty) [inline pkg/controller/namespace/deletion/namespaced_resources_deleter.go:452]: @soltysh I can't see a nil check for the `unstructuredList`. Should we handle this (maybe separate PR?)

File: pkg/controller/namespace/deletion/namespaced_resources_deleter.go
Comment (helayoty) [inline pkg/controller/namespace/deletion/namespaced_resources_deleter.go:467]:
```suggestion
		logger.V(5).Info("Namespace controller - deleteAllContentForGroupVersionResource - estimate is present -", "namespace", namespace, "groupVersionResource", gvr, "finalizers", finalizersToNumRemaining)
```

File: pkg/controller/namespace/deletion/namespaced_resources_deleter.go
Comment (helayoty) [inline pkg/controller/namespace/deletion/namespaced_resources_deleter.go:477]:
```suggestion
		logger.V(5).Info("Namespace controller - deleteAllContentForGroupVersionResource - items remaining with finalizers -", "namespace", namespace, "groupVersionResource", gvr, "finalizers", finalizersToNumRemaining)
```

File: pkg/controller/namespace/namespace_controller.go
Comment (helayoty) [inline pkg/controller/namespace/namespace_controller.go:154]:
```suggestion
			klog.FromContext(ctx).V(4).Info("Content remaining in namespace", key, "waitSeconds", t)
```

**Review by helayoty — COMMENTED** [review]
Thank you @yangjunmyfm192085 for your PR. Left nit comments to have as minimum changes in the logs as possible.

@soltysh

File: pkg/controller/namespace/deletion/namespaced_resources_deleter.go
Comment (yangjunmyfm192085) [reply pkg/controller/namespace/deletion/namespaced_resources_deleter.go:182]: thanks for you suggestion. but I think we don't need dots and extra spaces

File: pkg/controller/namespace/deletion/namespaced_resources_deleter.go
Comment (yangjunmyfm192085) [reply pkg/controller/namespace/deletion/namespaced_resources_deleter.go:315]: why do we need - here?

File: pkg/controller/namespace/deletion/namespaced_resources_deleter.go
Comment (yangjunmyfm192085) [reply pkg/controller/namespace/deletion/namespaced_resources_deleter.go:311]: why do we need -?

File: pkg/controller/namespace/deletion/namespaced_resources_deleter.go
Comment (yangjunmyfm192085) [reply pkg/controller/namespace/deletion/namespaced_resources_deleter.go:111]: same doubt

Comment (yangjunmyfm192085) [reply pkg/controller/namespace/deletion/namespaced_resources_deleter.go:111]: same doubt

File: pkg/controller/namespace/deletion/namespaced_resources_deleter.go
Comment (yangjunmyfm192085) [reply pkg/controller/namespace/deletion/namespaced_resources_deleter.go:336]: same doubt

File: pkg/controller/namespace/deletion/namespaced_resources_deleter.go
Comment (yangjunmyfm192085) [reply pkg/controller/namespace/deletion/namespaced_resources_deleter.go:340]: I think we don't need this `-` indeed

File: pkg/controller/namespace/deletion/namespaced_resources_deleter.go
Comment (yangjunmyfm192085) [reply pkg/controller/namespace/deletion/namespaced_resources_deleter.go:356]: I think we don't need this - indeed

File: pkg/controller/namespace/deletion/namespaced_resources_deleter.go
Comment (yangjunmyfm192085) [reply pkg/controller/namespace/deletion/namespaced_resources_deleter.go:371]: I think we don't need this - indeed

File: pkg/controller/namespace/deletion/namespaced_resources_deleter.go
Comment (yangjunmyfm192085) [reply pkg/controller/namespace/deletion/namespaced_resources_deleter.go:380]: I think we don't need this - indeed

File: pkg/controller/namespace/deletion/namespaced_resources_deleter.go
Comment (yangjunmyfm192085) [reply pkg/controller/namespace/deletion/namespaced_resources_deleter.go:417]: I think we don't need this - indeed

File: pkg/controller/namespace/namespace_controller.go
Comment (yangjunmyfm192085) [reply pkg/controller/namespace/namespace_controller.go:177]: thank you so much for your suggestion. Log structure can refer to
https://github.com/kubernetes/community/blob/master/contributors/devel/sig-instrumentation/migration-to-structured-logging.md
Thanks again for your advice

**Comment by yangjunmyfm192085** [comment]
@helayoty thank you so much for your suggestion. Log structure can refer to
https://github.com/kubernetes/community/blob/master/contributors/devel/sig-instrumentation/migration-to-structured-logging.md
Thanks again for your advice

File: pkg/controller/namespace/deletion/namespaced_resources_deleter.go
Comment (helayoty) [reply pkg/controller/namespace/deletion/namespaced_resources_deleter.go:315]: To have the same consistent log as the old ones. This applies to all other comments as well.

Comment (yangjunmyfm192085) [reply pkg/controller/namespace/deletion/namespaced_resources_deleter.go:315]: I don't think it needs to be consistent with the old logs. After the log is structured, it is already in key value format.

File: pkg/controller/namespace/deletion/namespaced_resources_deleter.go
Comment (pohly) [reply pkg/controller/namespace/deletion/namespaced_resources_deleter.go:311]: Yes, let's not add that. It's similar to "end with a period".

The [guidelines](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-instrumentation/migration-to-structured-logging.md#remove-string-formatting-from-log-message) for log messages are

* Start from a capital letter.
* Do not end the message with a period
* Use active voice. Use complete sentences when there is an acting subject ("A could not do B") or omit the subject if the subject would be the program itself ("Could not do B").
* Use past tense ("Could not delete B" instead of "Cannot delete B")
* When referring to an object, state what type of object it is. ("Deleted pod" instead of "Deleted")

Comment (pohly) [reply pkg/controller/namespace/deletion/namespaced_resources_deleter.go:311]: I find `groupVersionResource` too verbose: the item is a "resource", identified by a group/version/resource.

File: cmd/kube-controller-manager/app/core.go
Comment (pohly) [reply cmd/kube-controller-manager/app/core.go:459]: Or better, "namespace" as in https://github.com/kubernetes/kubernetes/blob/ca7804fe1374de9621d6806d81a91d82b8f42009/cmd/kube-controller-manager/app/controllermanager.go#L440

It doesn't matter, though, because I'll move the entire name handling out of the controllers into cmd/kube-controller-manager.

File: pkg/controller/namespace/deletion/namespaced_resources_deleter.go
Comment (pohly) [reply pkg/controller/namespace/deletion/namespaced_resources_deleter.go:422]: Probably not. The error is being handled automatically and not something that an admin needs to deal with.

File: pkg/controller/namespace/deletion/namespaced_resources_deleter.go
Comment (pohly) [reply pkg/controller/namespace/deletion/namespaced_resources_deleter.go:452]: It cannot be nil. That can only happen when an error is returned by `List` and that is checked.

**Review by pohly — APPROVED** [review]
/lgtm
/assign @soltysh 

For approval.

File: pkg/controller/namespace/deletion/namespaced_resources_deleter.go
Comment (soltysh) [reply pkg/controller/namespace/deletion/namespaced_resources_deleter.go:422]: Yeah, we're returning error to the caller, so Info(5) is good enough, if you want to dive deeper.

**Review by soltysh — APPROVED** [review]
/approve

**Pull request merged**
